### PR TITLE
text-layout v2

### PR DESCRIPTION
**Pull request by jmars**
Todos as discussed in the meeting 
- [x] Support scaling (font-forge)
- [x] Test with ios level font changes (**Jaye**)
- [x] Unit tests for text-flow package and some integrations
- [x] Change fontforge cli to create a js for font
- [x] Disable para layout for no inlines (**NATE**)
- [x] Add widths in all examples on readme
- [x] Add rendering integration example

### CODE REVIEW

File: packages/article-skeleton/src/article-skeleton.js [hunk 1]
```diff
@@ -210,6 +247,78 @@ class ArticleSkeleton extends Component {
...
+    const { width } = this.state;
+    if (!this.state.content) {
+      return null
+    }
+    const textFlow = this.state.content
+    return (
+      <AdComposer adConfig={adConfig}>
+        <Responsive>
+          <View style={styles.articleContainer}>
+            <ScrollView>
```
Comment (nehasri89) [inline packages/article-skeleton/src/article-skeleton.js]: Is there a reason we are not using Flatlist, besides we are not maing use of `onViewableItemsChanged` method which is used for analytics

File: packages/markup/src/markup-flow.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        children: renderedChildren
+      })
+    };
+  },
+  paragraph(key, attributes, renderedChildren) {
+    return {
+      element: new Text.Text({
+        markup: renderedChildren,
+        lineHeight: 30,
+        width: 600,
```
Comment (nehasri89) [inline packages/markup/src/markup-flow.js]: styles should come from a separate object

File: packages/text-flow/__tests__/shared.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import TestRenderer from "react-test-renderer";
+import TextFlow from "../src/text-flow";
+
+export default () => {
+  it("renders correctly", () => {
+    const testInstance = TestRenderer.create(
```
Comment (nehasri89) [inline packages/text-flow/__tests__/shared.js]: needs more test

**Comment by aronshamash** [comment]
@jmars Any chance you could add a description to the PR ?

File: packages/text-flow/src/Layout/BlockAlign.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export default class BlockAlign {
```
Comment (nehasri89) [inline packages/text-flow/src/Layout/BlockAlign.js]: should just be enum

File: packages/text-flow/src/Text/Box.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export default class Box {
+  type = "box"
```
Comment (nehasri89) [inline packages/text-flow/src/Text/Box.js]: const?

File: packages/text-flow/src/Text/Character.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
+import Positioned from "../Layout/Positioned";
+import Case from './Case';
+import FontLoader from './FontLoader';
+
+export default class Character extends Positioned {
+  character = '';
+  characterCode = null;
+  font = null;
+  tracking = null;
+  characterCase = null;
```
Comment (nehasri89) [inline packages/text-flow/src/Text/Character.js]: let or const?

File: packages/text-flow/README.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
+# TextFlow
+
+![Class Diagram](classes.svg)
+
+TextFlow is a package for easily doing text heavy layouts on platforms that don't have a typesetting engine (like React Native). It allows you to flow text around block elements, measure text sizing without platform support and fit text into arbitary shapes. It also has markup support for styling such as different fonts, weights, kerning, tracking and colours.
+
+## Usage
+### Measuring Text
+```javascript
+import { Text as FlowText, Markup } from "text-flow"
```
Comment (nathggns) [inline packages/text-flow/README.md]: Is this supposed to be `@times-components/text-flow`?

File: packages/article-skeleton/__tests__/common/dropcap-util.test.js [hunk 8]
```diff
@@ -18,17 +18,10 @@ const child = {
...
-        value: "A"
-      },
-      children: [],
-      name: "dropCap"
-    },
     {
       attributes: {
         value:
-          "s I follow Chris Reynolds Gordon down the first f…fty Shades of Grey but, he says, “I’m living it.”"
+          "As I follow Chris Reynolds Gordon down the first f…fty Shades of Grey but, he says, “I’m living it.”"
```
Comment (tuncaulubilge) [inline packages/article-skeleton/__tests__/common/dropcap-util.test.js]: Something wrong on this test?

File: packages/article-skeleton/__tests__/shared.native.js [hunk 9]
```diff
@@ -92,6 +97,8 @@ export default () => {
           renderArticle({ onLinkPress })
         );
 
+        expect(testInstance).toMatchSnapshot();
```
Comment (tuncaulubilge) [inline packages/article-skeleton/__tests__/shared.native.js]: nooo

File: packages/article-skeleton/src/article-skeleton.js [hunk 10]
```diff
@@ -28,32 +29,25 @@ const viewabilityConfig = {
   waitForInteraction: false
 };
 
+const convertStyles = ({ font, size }) => ({
+  fontFamily: font,
+  fontSize: size,
+  fontStyle: font.includes("Italic") ? "italic" : null,
+  fontWeight: font.includes("Bold") ? "bold" : null
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: Not sure if we need to add an extra bold in here. The font itself is already bold. Same for italic

File: packages/article-skeleton/src/article-skeleton.js [hunk 11]
```diff
@@ -72,13 +66,79 @@ const renderRow = (
...
+  return block.getComponent(style => (
+    <View style={{ height: block.measuredHeight }}>
+      {block.block.children.map(line =>
+        line.idealSpans.map(span => {
+          if (span.href) {
+            return (
+              <Text
+                selectable
+                style={{
+                  left: span.x,
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: Do we need the absolute x pos?

Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: `...style`

File: packages/article-skeleton/src/article-skeleton.js [hunk 12]
```diff
@@ -100,36 +164,42 @@ class ArticleSkeleton extends Component {
...
+      }
+      return {
+        data,
+        type: "articleBodyRow"
+      };
+    });
+  }
+
+  layout() {
     const {
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: We can try to split this function into smaller pieces for readability

File: packages/article-skeleton/src/article-skeleton.js [hunk 13]
```diff
@@ -142,14 +212,110 @@ class ArticleSkeleton extends Component {
...
+          const { width } = this.state;
+
+          const onLayout = () => {
+            if (block.prevHeight !== block.height) {
+              this.layout();
+            }
+            block.prevHeight = block.height;
+          };
+
+          // @todo Is there a better way to get this
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: todo?

File: packages/article-skeleton/src/article-skeleton.js [hunk 14]
```diff
@@ -142,14 +212,110 @@ class ArticleSkeleton extends Component {
...
+  }
+
+  renderInlineBlock(block) {
+    return (
+      <ResponsiveContext.Consumer>
+        {() => {
+          const { width } = this.state;
+
+          const onLayout = () => {
+            if (block.prevHeight !== block.height) {
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: Make sure this doesn't cause an infinite loop

File: packages/article-skeleton/src/article-skeleton.js [hunk 15]
```diff
@@ -142,14 +212,110 @@ class ArticleSkeleton extends Component {
...
+    }
+
+    const { fontScale } = Dimensions.get("window");
+
+    const textFlow = new Layout.TextFlow({
+      flow: rows.map(rowData =>
+        ArticleRowFlow({
+          content: rowData,
+          fontScale,
+          interactiveConfig,
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/article-skeleton.js]: onImagePress should also be here

File: packages/article-skeleton/src/dropcap-util.js [hunk 16]
```diff
@@ -33,18 +34,23 @@ const insertDropcap = (firstTextChild, children) => {
...
     template &&
     templateWithDropCaps.includes(template) &&
     !isDropcapDisabled &&
-    name === "paragraph" &&
-    children.length > 0
+    children.length >= 1 &&
+    children[0].name === "paragraph"
   ) {
-    return { ...child, children: insertDropcap(children[0], children) };
+    const firstTextChild = children[0].children[0];
```
Comment (tuncaulubilge) [inline packages/article-skeleton/src/dropcap-util.js]: check children[0].children.length > 0

File: packages/markup/src/markup-flow.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    return {
+      element: new Markup.Italic({
+        children: renderedChildren
+      })
+    };
+  },
+  paragraph(key, attributes, renderedChildren) {
+    return {
+      element: new Text.Text({
+        font: "TimesDigitalW04-Regular",
```
Comment (tuncaulubilge) [inline packages/markup/src/markup-flow.js]: it should be getting from fontFactory

File: packages/styleguide/src/line-heights/mapping-base.js [hunk 18]
```diff
@@ -1,6 +1,6 @@
 const mapping = {
   body: {
-    bodyMobile: 26,
+    bodyMobile: 30,
```
Comment (tuncaulubilge) [inline packages/styleguide/src/line-heights/mapping-base.js]: Let's review this with Matt afterwards

File: packages/text-flow/README.md [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+Or the tests for all platforms can be run
+
+```
+yarn test:all
+```
+
+Visit the official
+
+<!-- Add the storybook link here. -->
```
Comment (tuncaulubilge) [inline packages/text-flow/README.md]: You can add the storybook link here

File: packages/text-flow/src/Layout/TextFlow.js [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+      if (child instanceof TextFlow) {
+        throw new Error("Nested text flows is not supported");
+      }
+      if (child instanceof InlineBlock) {
+        let grabbed = i + 1;
+        let next = this.flow[grabbed];
+        child.children = [];
+
+        if (!(next instanceof Text)) {
+          i += 1
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Layout/TextFlow.js]: We shouldn't skip the InlineBlock, but just display it as a regular block

File: packages/text-flow/src/Layout/TextFlow.js [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+          next = this.flow[grabbed];
+          if (pulledLines >= elNumLines || !next) {
+            break;
+          }
+        }
+        i = grabbed;
+        this.block.measuredHeight += child.height;
+        continue
+      }
+      if (child instanceof Block) {
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Layout/TextFlow.js]: This is same as text

File: packages/text-flow/src/Markup/Bold.js [hunk 22]
```diff
@@ -0,0 +1,8 @@
+import StyledText from "./Styled";
+
+export default class BoldMarkup extends StyledText {
+  constructor(props = {}) {
+    super(props);
+    this.style.font = "TimesDigitalW04-Bold";
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Markup/Bold.js]: Let's try to avoid hardcoding this

File: packages/text-flow/src/Markup/Italic.js [hunk 23]
```diff
@@ -0,0 +1,8 @@
+import StyledText from "./Styled";
+
+export default class ItalicMarkup extends StyledText {
+  constructor(props = {}) {
+    super(props);
+    this.style.font = "TimesDigitalW04-Italic";
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Markup/Italic.js]: Let's try to avoid hardcoding this

File: packages/text-flow/src/Markup/Link.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable no-param-reassign */
+import StyledText from "./Styled";
+
+export default class LinkText extends StyledText {
+  href = null;
+
+  constructor(props = {}) {
+    super(props);
+    this.href = props.href;
+    this.style.font = "TimesDigitalW04-Bold";
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Markup/Link.js]: bold?

File: packages/text-flow/src/Markup/String.js [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+export default class MarkupString {
+  text = "";
+
+  constructor(text = "") {
+    this.text = text;
+  }
+
+  characters(textStyle) {
+    return this.text
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Markup/String.js]: Do we need this? We should use string as it is

File: packages/text-flow/src/Markup/Styled.js [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+export default class StyledText extends Markup {
+  style = new TextStyle();
+
+  characters(textStyle) {
+    if (!this.style.size) this.style.size = textStyle.size;
+    if (!this.style.font) this.style.font = textStyle.font;
+    if (!this.style.tracking) this.style.tracking = textStyle.tracking;
+    if (!this.style.characterCase)
+      this.style.characterCase = textStyle.characterCase;
+    if (!this.style.fillColor) this.style.fillColor = textStyle.fillColor;
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Markup/Styled.js]: Do we need colours in here?

File: packages/text-flow/src/Text/fonts/timesdigitalw04_bold.js [hunk 27]
```diff
@@ -0,0 +1,593 @@
+export default `0|missing|489
```
Comment (tuncaulubilge) [inline packages/text-flow/src/Text/fonts/timesdigitalw04_bold.js]: We need to ignore these

File: yarn.lock [hunk 28]
```diff
@@ -1824,7 +1824,7 @@
     react-split-pane "^0.1.77"
     react-treebeard "^2.1.0"
 
-"@thetimes/jest-lint@*":
+"@thetimes/jest-lint@*", "@thetimes/jest-lint@1.4.0":
```
[the remark this answers is not in the document]
Comment (jmars) [reply yarn.lock]: fixed with a yarn install

File: packages/text-flow/README.md [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+const text = new FlowText.Text({
+  font: "TimesModern-Regular",
+  width: 660,
+  size: 18,
+  lineHeight: 30,
+  markup: [new Markup.MarkupString("Hello World!")]
+});
+
+const width = text.measuredWidth;
+const Height = text.measuredHeight;
```
Comment (aronshamash) [inline packages/text-flow/README.md]: lowercase `h`

**Pull request merged**